ArduCopter/mode_autorotate.cpp

```diff
@@ -0,0 +1,338 @@
+/* -----------------------------------------------------------------------------------------
+    This is currently a SITL only function until the project is complete.
+    To trial this in SITL you will need to use Real Flight 8.
+    Instructions for how to set this up in SITL can be found here:
+    https://discuss.ardupilot.org/t/autonomous-autorotation-gsoc-project-blog/42139
+ -----------------------------------------------------------------------------------------*/
+
+#include "Copter.h"
+#include <AC_Autorotation/AC_Autorotation.h>
+#include "mode.h"
+
+#include <utility>
+
+#if MODE_AUTOROTATE_ENABLED == ENABLED
+
+#define AUTOROTATE_ENTRY_TIME          2.0f    // (s) number of seconds that the entry phase operates for
+#define BAILOUT_MOTOR_RAMP_TIME        1.0f    // (s) time set on bailout ramp up timer for motors - See AC_MotorsHeli_Single
+#define HEAD_SPEED_TARGET_RATIO        1.0f    // Normalised target main rotor head speed (unit: -)
+
+bool ModeAutorotate::init(bool ignore_checks)
+{
+#if FRAME_CONFIG != HELI_FRAME
+    // Only allow trad heli to use autorotation mode
+    return false;
+#endif
+
+    // Check that mode is enabled
+    if (!g2.arot.is_enable()) {
+        gcs().send_text(MAV_SEVERITY_INFO, "Autorotation Mode Change Fail: Mode Not Enabled");
+        return false;
+    }
+
+    // Check that interlock is disengaged
+    if (motors->get_interlock()) {
+        gcs().send_text(MAV_SEVERITY_INFO, "Autorotation Mode Change Fail: Interlock Engaged");
+        return false;
+    }
+
+    // Initialise head speed/collective controller
+    // This must be done before RPM value is fetched
+    g2.arot.init_hs_controller();
+
+    // Retrive rpm and start rpm sensor health checks
+    _initial_rpm = g2.arot.get_rpm(true);
+
+    // Display message 
+    gcs().send_text(MAV_SEVERITY_INFO, "Autorotation initiated");
+
+    // Retrieve parameter values from autorotation library
+    g2.arot.get_param_values(_param_head_speed_set_point, _param_accel_max, _param_target_fwd_speed, _param_col_entry_cutoff_freq, _param_col_glide_cutoff_freq, _param_bail_time);
```

Comment:
I wonder why we maintain copies of these param values in the mode instead of adding individual accessors to the arot object and then use the values directly in the mode class?

Reply:
I think I was mistaken as to why I did this in the first place.  I have now done as you suggest.